src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs

```diff
@@ -141,6 +136,14 @@ private async Task ProduceTagsAsync(TaggerContext<TTag> context, DocumentSnapsho
             var cancellationToken = context.CancellationToken;
             var workspace = document.Project.Solution.Workspace;
 
+            var diagnosticCacheService = workspace.Services.GetService<IDiagnosticCacheService>();
+            if (diagnosticCacheService != null)
+            {
+                // During solution load. if there's any cached diagnostics available,
+                // this service will (asynchronously) push updates to IDiagnosticService.
```

Comment:
having hte tagger push to the DiagService seems like a *very* *very* strange data flow arrangement.  Tagging @mavasani .  Iwould not have expected this.  I would prefer we treat the IDiagService as the source of truth, with all our features being driven by the results it produces.

Reply:
We only need something that's aware of opened document as trigger, it doesn't need to be tagger itself. It can't be solution crawler though, since we are operating before it started. Maybe raw workspace events?